broker.go

```diff
@@ -0,0 +1,645 @@
+	// sync for a specific peer.
+	Cid cid.Cid
+	// PeerID identifies the peer this SyncFinished event pertains to.
+	PeerID peer.ID
+}
+
+// handler holds state that is specific to a peer
+type handler struct {
+	dtManager dt.Manager
+	// distEvents is used to communicate a syncDone event back to the Broker
+	// for distribution to OnSyncFinished readers.
+	distEvents chan<- SyncFinished
+	latestSync ipld.Link
+	msgChan    chan cid.Cid
+	// peerID is the ID of the peer this handler is responsible for.
+	peerID peer.ID
+	// putSyncDone is a function supplied by the Broker for the handler to give
+	// Broker the channel to send sync done notification to.
+	putSyncDone func(cid.Cid, chan<- struct{})
+	// syncMutex serializes the handling of individual syncs
+	syncMutex sync.Mutex
+}
+
+// NewBroker creates a new Broker that process pubsub messages.
+func NewBroker(host host.Host, ds datastore.Batching, lsys ipld.LinkSystem, topic string, dss ipld.Node, options ...Option) (*Broker, error) {
+	cfg := config{
+		addrTTL: defaultAddrTTL,
+	}
+	err := cfg.apply(options)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	ctx, cancelAll := context.WithCancel(context.Background())
+
+	pubsubTopic := cfg.topic
+	if pubsubTopic == nil {
+		pubsubTopic, err = makePubsub(ctx, host, topic)
+		if err != nil {
+			cancelAll()
+			return nil, err
+		}
+	}
+	psub, err := pubsubTopic.Subscribe()
+	if err != nil {
+		cancelAll()
+		return nil, err
+	}
+
+	dt, gs, tmpDir, err := makeDataTransfer(ctx, host, ds, lsys, cfg.dtManager)
+	if err != nil {
+		cancelAll()
+		return nil, err
+	}
+
+	lb := &Broker{
+		dtManager:  dt,
+		dss:        dss,
+		host:       host,
+		gsExchange: gs,
+		tmpDir:     tmpDir,
+
+		addrTTL:   cfg.addrTTL,
+		psub:      psub,
+		topic:     pubsubTopic,
+		closing:   make(chan struct{}),
+		cancelAll: cancelAll,
+
+		allowPeer: cfg.allowPeer,
+		handlers:  make(map[peer.ID]*handler),
+		inEvents:  make(chan SyncFinished, 1),
+	}
+
+	lb.unsubEvents = dt.SubscribeToEvents(lb.onEvent)
+
+	// Start watcher to read pubsub messages.
+	go lb.watch(ctx)
+	// Start distributor to send SyncFinished messages to interested parties.
+	go lb.distributeEvents()
+
+	return lb, nil
+}
+
+// GetLatestSync returns the latest synced CID for the specified peer. If there
+// is not handler for the peer, then nil is returned.  This does not mean that
+// no data is synced with that peer, it means that the broker does not know
+// about it.  Calling Sync() first may be necessary.
+func (lb *Broker) GetLatestSync(peerID peer.ID) ipld.Link {
+	lb.handlersMutex.Lock()
+	hnd, ok := lb.handlers[peerID]
+	if !ok {
+		lb.handlersMutex.Unlock()
+		return nil
+	}
+	lb.handlersMutex.Unlock()
+
+	return hnd.getLatestSync()
+}
+
+// SetLatestSync sets the latest synced CID for a specified peer.  If there is
+// no handler for the peer, then one is created without consulting any
+// AllowPeerFunc.
+func (lb *Broker) SetLatestSync(peerID peer.ID, latestSync cid.Cid) error {
```

Comment:
What do you think abouut this func returning an error if you pass in `latestSync == cid.Undef`? This way the caller gets an error rather this it silently doing nothing.

Reply:
I was thinking that passing `cid.Undef` provided a way to create a handler without actually doing a sync.  That might be useful for bypassing the AllowPeer without doing a sync.

However, I cannot really think of a use case where it would not be preferable to update the AllowPeer instead or just do an explicit sync, so passing `cid.Undef` is more likely the result of an error and should therefore be treated as such.